src/k8s/pkg/k8sd/controllers/node_configuration.go

```diff
@@ -0,0 +1,79 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/canonical/k8s/pkg/k8sd/types"
+	"github.com/canonical/k8s/pkg/snap"
+	snaputil "github.com/canonical/k8s/pkg/snap/util"
+	"github.com/canonical/k8s/pkg/utils/k8s"
+	"github.com/mitchellh/mapstructure"
+	v1 "k8s.io/api/core/v1"
+)
+
+type NodeConfigurationController struct {
+}
+
+func NewNodeConfigurationController() *NodeConfigurationController {
+	return &NodeConfigurationController{}
+}
+
+func (c *NodeConfigurationController) Run(ctx context.Context, snap snap.Snap) error {
```

Comment:
Make the snap and the k8s client a part of the controller itself, and consider adding a unit test (using a mock snap and client)

Reply:
Otherwise we create clients in a loop, which works but will eventually lead to fd leaks

Reply:
If we need to recreate the client (e.g. because config files are not yet ready, consider making this a function that creates the client